README.md
# credit_card_number_validator

A dart package to check if card number is valid and also returns card type as well.

# Installing
1. Add dependency to `pubspec.yaml`

    *Get the current version from the 'Installing' tab*
```
dependencies:
    credit_card_number_validator: *current-version*
```

2. Import the package
```
import 'package:credit_card_number_validator/credit_card_number_validator.dart';
```

#  Usage
A basic example

    Map<String, dynamic> cardData = CreditCardValidator.getCard(cardNumberString);
    String cardType = cardData[CreditCardValidator.cardType];
    bool isValid = cardData[CreditCardValidator.isValidCard];

```
# Features
* Supported cards:
  * Visa
  * Mastercard
  * American Express
  * Discover
  * Diners Club
  * JCB


# Author
Preeth Prathapan - *Creator and repo owner* - [Github Profile](https://github.com/preethzcodez)

# License
This project is licensed under the MIT License - see the [LICENSE file](LICENSE) for more details


CHANGELOG.md
## [1.0.4] - July 23, 2019
* Fixed Issue With MasterCard & JCB

## [1.0.3] - July 23, 2019
* Updated README.md

## [1.0.2] - July 23, 2019
* Stable Build

## [1.0.1] - July 23, 2019
* Updated README.md

## [1.0.0] - July 23, 2019
* Added Documentation

## [0.0.2] - July 22, 2019
* Updated README.md

## [0.0.1] - July 22, 2019
* Initial release


test/credit_card_validator_test.dart

import 'package:credit_card_number_validator/credit_card_number_validator.dart';
import 'package:flutter_test/flutter_test.dart';

void main() {
  test("Expect a VISA card", () {
    Map<String, dynamic> cardData =
        CreditCardValidator.getCard("4111111111111111");
    expect(cardData[CreditCardValidator.cardType], "VISA");
    expect(cardData[CreditCardValidator.isValidCard], true);
  });
}


example/example/test/credit_card_validator_test.dart
import 'package:credit_card_number_validator/credit_card_number_validator.dart';
import 'package:flutter_test/flutter_test.dart';

void main() {
  test("Expect a VISA card", () {
    Map<String, dynamic> cardData =
        CreditCardValidator.getCard("4111111111111111");
    expect(cardData[CreditCardValidator.cardType], "VISA");
    expect(cardData[CreditCardValidator.isValidCard], true);
  });
}


example/main.dart
import 'package:credit_card_number_validator/credit_card_number_validator.dart';
import 'package:flutter/material.dart';

void main() => runApp(MyApp());

class MyApp extends StatefulWidget {
  @override
  _MyApp createState() => _MyApp();
}

class _MyApp extends State<MyApp> {
  /// Card Number Controller
  TextEditingController _cardNumberController = TextEditingController();

  // Declare Variables To Store Card Type and Validity
  String cardType;
  bool isValid = false;

  // Initial State
  @override
  void initState() {
    // TODO: implement initState
    super.initState();

    // Set A Sample Card Value - VISA Card
    _cardNumberController.text = "4111111111111111";
  }

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
        home: Scaffold(
      appBar: AppBar(
        title: Text('Credit Card Validator Demo'),
      ),
      body: Center(
        child: Container(
            child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            Text(
              'ENTER CARD NUMBER',
              style: TextStyle(fontSize: 14.0, fontWeight: FontWeight.w800),
            ),
            Padding(
              padding: EdgeInsets.all(16.0),
              child: TextFormField(
                  keyboardType: TextInputType.number,
                  controller: _cardNumberController,
                  decoration: InputDecoration(
                      hintText: 'Card Number',
                      hintStyle: TextStyle(color: Color(0xFFCCCCCC)),
                      contentPadding: new EdgeInsets.symmetric(
                          vertical: 14.0, horizontal: 7.0),
                      focusedBorder: const OutlineInputBorder(
                        borderSide:
                            const BorderSide(color: Colors.grey, width: 0.0),
                      ),
                      enabledBorder: const OutlineInputBorder(
                        borderSide:
                            const BorderSide(color: Colors.grey, width: 0.0),
                      ),
                      border: const OutlineInputBorder(
                        borderSide:
                            const BorderSide(color: Colors.grey, width: 0.0),
                      ))),
            ),
            Container(
              height: 55,
              child: RaisedButton(
                color: Colors.blue,
                shape: RoundedRectangleBorder(
                    borderRadius: BorderRadius.all(Radius.circular(5))),
                child: Text('Get Card Data',
                    style: TextStyle(color: Colors.white, fontSize: 18.0)),
                onPressed: () {
                  // Get Card Type and Validity Data As Map - @param Card Number
                  Map<String, dynamic> cardData =
                      CreditCardValidator.getCard(_cardNumberController.text);
                  setState(() {
                    // Set Card Type and Validity
                    cardType = cardData[CreditCardValidator.cardType];
                    isValid = cardData[CreditCardValidator.isValidCard];
                  });
                },
              ),
            ),
            Padding(
              padding: EdgeInsets.all(20),
              child: cardType != null
                  // Display Card Type and Validity
                  ? Text('Card Type : $cardType \nCard Number Valid: $isValid',
                      style: TextStyle(
                          color: isValid ? Colors.green : Colors.red,
                          fontSize: 14.0,
                          fontWeight: FontWeight.w800))
                  : Text(' \n '),
            )
          ],
        )),
      ), // This trailing comma makes auto-formatting nicer for build methods.
    ));
  }
}


lib/credit_card_validator.dart
library credit_card_validator;

class CreditCardValidator {
  static const String _VISA = "^4[0-9]{12}(?:[0-9]{3})?\$";
  static const String _MASTERCARD =
      "^(?:5[1-5][0-9]{2}|222[1-9]|22[3-9][0-9]|2[3-6][0-9]{2}|27[01][0-9]|2720)[0-9]{12}\$";
  static const String _DISCOVER = "^6(?:011|5[0-9]{2})[0-9]{12}\$";
  static const String _AMEX = "^3[47][0-9]{13}\$";
  static const String _DINERS = "^3(?:0[0-5]|[68][0-9])[0-9]{11}\$";
  static const String _JCB = "^(?:2131|1800|35\d{3})\d{11}\$";

  static const String cardType = "CardType";
  static const String isValidCard = "IsValidCard";

  static const List<String> regexList = [
    _VISA,
    _MASTERCARD,
    _DISCOVER,
    _AMEX,
    _DINERS,
    _JCB
  ];

  static const List<String> _cardTypesList = [
    "VISA",
    "MASTERCARD",
    "DISCOVER",
    "AMEX",
    "DINERS",
    "JCB"
  ];

  static RegExp _regex = new RegExp("^(?:(?<visa>4[0-9]{12}(?:[0-9]{3})?)|" +
      "(?<mastercard>5[1-5][0-9]{2}|222[1-9]|22[3-9][0-9]|2[3-6][0-9]{2}|27[01][0-9]|2720)[0-9]{12}" +
      "(?<discover>6(?:011|5[0-9]{2})[0-9]{12})|" +
      "(?<amex>3[47][0-9]{13})|" +
      "(?<diners>3(?:0[0-5]|[68][0-9])[0-9]{11})|" +
      "(?<jcb>(?:2131|1800|35\d{3})\d{11}))\$");

  static Map<String, dynamic> getCard(String cardNumber) {
    String _cardType = _getCardType(cardNumber);
    bool isValid = _isCardValid(_cardType, cardNumber);

    Map<String, dynamic> cardData = new Map();
    cardData.putIfAbsent(cardType, () => _cardType);
    cardData.putIfAbsent(isValidCard, () => isValid);
    return cardData;
  }

  static String _getCardType(String cardNumber) {
    if (_regex.hasMatch(cardNumber)) {
      for (int i = 0; i < regexList.length; i++) {
        RegExp regExp = new RegExp(regexList[i]);
        if (regExp.hasMatch(cardNumber)) {
          return _cardTypesList[i];
        }
      }
    }
    return "UNKNOWN";
  }

  static bool _isCardValid(String cardType, String ccNumber) {
    if (cardType == "UNKNOWN" || cardType.trim().length == 0) return false;

    int sum = 0;
    bool alternate = false;
    for (int i = ccNumber.length - 1; i >= 0; i--) {
      int n = int.parse(ccNumber.substring(i, i + 1));
      if (alternate) {
        n *= 2;
        if (n > 9) {
          n = (n % 10) + 1;
        }
      }
      sum += n;
      alternate = !alternate;
    }
    return (sum % 10 == 0);
  }
}


lib/credit_card_number_validator.dart
library credit_card_number_validator;

class CreditCardValidator {
  /// Set RegEx For All Cards
  static const String _VISA = "^4[0-9]{12}(?:[0-9]{3})?\$";
  static const String _MASTERCARD =
      "^(?:5[1-5][0-9]{2}|222[1-9]|22[3-9][0-9]|2[3-6][0-9]{2}|27[01][0-9]|2720)[0-9]{12}\$";
  static const String _DISCOVER = "^6(?:011|5[0-9]{2})[0-9]{12}\$";
  static const String _AMEX = "^3[47][0-9]{13}\$";
  static const String _DINERS = "^3(?:0[0-5]|[68][0-9])[0-9]{11}\$";
  static const String _JCB = "^(?:2131|2100|1800|3\\d{4})\\d{11}\$";

  /// Set Map Key
  static const String cardType = "CardType";
  static const String isValidCard = "IsValidCard";

  /// Add RegEx Cards Into List
  static const List<String> regexList = [
    _VISA,
    _MASTERCARD,
    _DISCOVER,
    _AMEX,
    _DINERS,
    _JCB
  ];

  /// Set Card Types List
  static const List<String> _cardTypesList = [
    "VISA",
    "MASTERCARD",
    "DISCOVER",
    "AMEX",
    "DINERS",
    "JCB"
  ];

  /// Combined RegEx
  static RegExp _regex = new RegExp("^(?:(?<visa>4[0-9]{12}(?:[0-9]{3})?)|" +
      "(?<mastercard>5[1-5][0-9]{2}|222[1-9]|22[3-9][0-9]|2[3-6][0-9]{2}|27[01][0-9]|2720)[0-9]{12}|" +
      "(?<discover>6(?:011|5[0-9]{2})[0-9]{12})|" +
      "(?<amex>3[47][0-9]{13})|" +
      "(?<diners>3(?:0[0-5]|[68][0-9])[0-9]{11})|" +
      "(?<jcb>(?:2131|2100|1800|3\\d{4})\\d{11}))\$");

  // Get Card Data - Card Type and isValid Number as Map
  static Map<String, dynamic> getCard(String cardNumber) {
    String _cardType = _getCardType(cardNumber);
    bool isValid = _isCardValid(_cardType, cardNumber);

    Map<String, dynamic> cardData = new Map();
    cardData.putIfAbsent(cardType, () => _cardType);
    cardData.putIfAbsent(isValidCard, () => isValid);
    return cardData;
  }

  // Get Card Type Based On RegEx
  static String _getCardType(String cardNumber) {
    if (_regex.hasMatch(cardNumber)) {
      for (int i = 0; i < regexList.length; i++) {
        RegExp regExp = new RegExp(regexList[i]);
        if (regExp.hasMatch(cardNumber)) {
          return _cardTypesList[i];
        }
      }
    }
    return "UNKNOWN";
  }

  // Check Card Number Validity - Luhn Algorithm
  static bool _isCardValid(String cardType, String ccNumber) {
    if (cardType == "UNKNOWN" || cardType.trim().length == 0) return false;

    int sum = 0;
    bool alternate = false;
    for (int i = ccNumber.length - 1; i >= 0; i--) {
      int n = int.parse(ccNumber.substring(i, i + 1));
      if (alternate) {
        n *= 2;
        if (n > 9) {
          n = (n % 10) + 1;
        }
      }
      sum += n;
      alternate = !alternate;
    }
    return (sum % 10 == 0);
  }
}
